### ISSUE TITLE
Crash after upgrade to v1.0.0

### ISSUE THREAD

**Issue by Blankwonder** [ngtcp2/ngtcp2#988]
Hi. I have been using the v0.16.0 version previously, which has always been very stable. After upgrading to the v1.0.0 version, users reported crashes, and the specific crash report is as follows:

```
Exception Type:  SIGSEGV
Exception Codes: SEGV_MAPERR at 0xb

0   ngtcp2_pkt_encode_datagram_frame (ngtcp2_pkt.c:2035)
1   ngtcp2_ppe_encode_frame (ngtcp2_ppe.c:95)
2   conn_ppe_write_frame_hd_log (ngtcp2_conn.c:1805)
3   conn_write_pkt (ngtcp2_conn.c:4031)
4   ngtcp2_conn_write_vmsg (ngtcp2_conn.c:12195)
5   ngtcp2_conn_writev_datagram_versioned (ngtcp2_conn.c:11895)
```

Please let me know if you need more information.

**Comment by tatsuhiro-t** [comment]
If you use NGTCP2_WRITE_DATAGRAM_FLAG_MORE, make sure that the return value from ngtcp2_conn_writev_datagram is properly handled.  The buffer given to the function should not change until it produces a packet or returns 0.

**Comment by Blankwonder** [comment]
NGTCP2_WRITE_DATAGRAM_FLAG_MORE is never used.

**Comment by Blankwonder** [comment]
Related code:

```
    ngtcp2_pkt_info pi = {0};
    ngtcp2_ssize nwrite;
    uint8_t buf[2000];
    ngtcp2_path_storage ps;
    size_t datavcnt = 1;

    int paccepted;
    ngtcp2_path_storage_zero(&ps);

    ngtcp2_vec datav;

    datav.base = (uint8_t *)data.bytes;
    datav.len = data.length;

    uint32_t flags = 0;

    nwrite = ngtcp2_conn_writev_datagram(conn, &ps.path, &pi, buf, sizeof(buf),
                                         &paccepted, flags, 0, &datav, datavcnt, SGQUICTimestamp());
```

**Comment by tatsuhiro-t** [comment]
Thank you.  Are you using NGTCP2_WRITE_STREAM_FLAG_MORE flag with ngtcp2_conn_writev_stream?  The flag name is different, but the underlying machinery is shared.

**Comment by Blankwonder** [comment]
Yes. I'm using NGTCP2_WRITE_STREAM_FLAG_MORE.

However, I have never observed this issue in version v0.16.0, but due to some modifications in the related code, it cannot be fully confirmed whether it is caused by upgrading to v1.0.0. I have already downgraded back to v0.16.0 to see if the crash will still occur.

Additionally, I guess you think that the buffer has been released when accessing the written buffer, causing access to invalid memory. But combined with the error report `SEGV_MAPERR at 0xb`, it suggests that there should be a problem with the buf->last parameter when ngtcp2_ppe_encode_frame (ngtcp2_ppe.c:95) calls ngtcp2_pkt_encode_datagram_frame. This should not be a problem caused by external data.

**Comment by tatsuhiro-t** [comment]
buf is initialized with the given external buffer that application gives to the ngtcp2 functions.  If buf is stack allocated, and ngtcp2_conn_writev_stream or ngtcp2_conn_writev_datagram do not fully produce a packet or there is no packet to be made, buf keeps dangling pointer.

**Comment by Blankwonder** [comment]
Yes, I understand, but in this case, should the SEGV_MAPERR error address be a random high address rather than 0xb?

**Comment by tatsuhiro-t** [comment]
After 0.16.0, I made some fixes in the related code so that NGTCP2_ERR_STREAM_DATA_BLOCKED is returned where it should.  The fix makes the code consistent to the function specification, but it could introduce the side effect in the existing code.

I think 0xb is the signal code for SIGSEGV, not an address.

**Comment by Blankwonder** [comment]
No. 0xb should be an address. Here are some similar reports from other crashes.

```
Exception Type:  SIGSEGV
Exception Codes: SEGV_MAPERR at 0x8
```

```
Exception Type:  SIGSEGV
Exception Codes: SEGV_MAPERR at 0x1e
```

```
Exception Type:  SIGSEGV
Exception Codes: SEGV_MAPERR at 0x21a1fdca2939
```

```
Exception Type:  SIGSEGV
Exception Codes: SEGV_MAPERR at 0x656e69676e655420
```

**Comment by tatsuhiro-t** [comment]
Thank you.  They seem to be coincidentally the same value.
In general, when the memory related problem occurs, the behavior of a program is undefined.   That means any data can be overwritten with the seemingly random data.  Not still fully understand why the value is there though.
